### PR TITLE
Added Smart driver features

### PR DESCRIPTION
**Pull request by HarshDaryani896**
This PR includes the implementation of Smart-driver features like Cluster Aware and Topology Aware Policies (with fallback).

Added the following 5 new connection properties: 

1. load_balance
2. topology_keys
3. yb_servers_refresh_interval
4. fallback_to_topology_keys_only
5. failed_host_reconnect_delay_secs

Testing:
Replicated existing example of pgx driver for rust smart driver, PR raised: https://github.com/yugabyte/driver-examples/pull/12

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,74 @@
...
+Note that, you would still need to specify `load_balance=true` to enable the topology-aware connection load balancing.
+
+```
+"postgresql://127.0.0.1:5433/yugabyte?user=yugabyte&password=yugabyte&load_balance=true&topology_keys=cloud1.datacenter1.rack1"
+```
+### Specifying fallback zones
+
+For topology-aware load balancing, you can now specify fallback placements too. This is not applicable for cluster-aware load balancing.
+Each placement value can be suffixed with a colon (`:`) followed by a preference value between 1 and 10.
+A preference value of `:1` means it is a primary placement. A preference value of `:2` means it is the first fallback placement and so on.If no preference value is provided, it is considered to be a primary placement (equivalent to one with preference value `:1`). Example given below.
```
Comment (ashetkar) [inline README.md]: nit: missing space after period
`fallback placement and so on.If` -> `fallback placement and so on. If`

File: postgres/src/config.rs [hunk 2]
```diff
@@ -81,6 +82,24 @@ use tokio_postgres::{Error, Socket};
 ///     `disable`, hosts and addresses will be tried in the order provided. If set to `random`, hosts will be tried
 ///     in a random order, and the IP addresses resolved from a hostname will also be tried in a random order. Defaults
 ///     to `disable`.
+/// * `load_balance` -  It expects true/false as its possible values. In YBClusterAwareDataSource load balancing is true
```
Comment (ashetkar) [inline postgres/src/config.rs]: Copy-paste miss: There is neither `YBClusterAwareDataSource` nor `DriverManager` class in rust :)

File: tokio-postgres/src/config.rs [hunk 3]
```diff
@@ -145,6 +160,24 @@ pub enum Host {
 ///     `disable`, hosts and addresses will be tried in the order provided. If set to `random`, hosts will be tried
 ///     in a random order, and the IP addresses resolved from a hostname will also be tried in a random order. Defaults
 ///     to `disable`.
+/// * `load_balance` -  It expects true/false as its possible values. In YBClusterAwareDataSource load balancing is true
```
Comment (ashetkar) [inline tokio-postgres/src/config.rs]: Remove references to `YBClusterAwareDataSource` and `DriverManager`

File: postgres/src/config.rs [hunk 4]
```diff
@@ -81,6 +82,24 @@ use tokio_postgres::{Error, Socket};
...
+///     to 'true'.
+/// * `topology_keys` - It takes a comma separated geo-location values. A single geo-location can be given as 'cloud.region.zone'.
+///     Multiple geo-locations too can be specified, separated by comma (,). Each placement value can be suffixed with a colon (:)
+///     followed by a preference value between 1 and 10. A preference value of :1 means it is a primary placement. A preference
+///     value of :2 means it is the first fallback placement and so on. If no preference value is provided, it is considered to
+///     be a primary placement (equivalent to one with preference value :1).
+/// * `yb_servers_refresh_interval` - Time interval, in seconds, between two attempts to refresh the information about cluster nodes.
+///     Default is 300. Valid values are integers between 0 and 600. Value 0 means refresh for each connection request. Any value
+///     outside this range is ignored and the default is used.
+/// * `fallback_to_topology_keys_only` - (default value: 5 seconds) The driver marks a server as failed with a timestamp, when it cannot
```
Comment (ashetkar) [inline postgres/src/config.rs]: The descriptions of `fallback_to_topology_keys_only` and `failed_host_reconnect_delay_secs` got interchanged. Please correct them

File: tokio-postgres/src/config.rs [hunk 5]
```diff
@@ -145,6 +160,24 @@ pub enum Host {
...
+///     to 'true'.
+/// * `topology_keys` - It takes a comma separated geo-location values. A single geo-location can be given as 'cloud.region.zone'.
+///     Multiple geo-locations too can be specified, separated by comma (,). Each placement value can be suffixed with a colon (:)
+///     followed by a preference value between 1 and 10. A preference value of :1 means it is a primary placement. A preference
+///     value of :2 means it is the first fallback placement and so on. If no preference value is provided, it is considered to
+///     be a primary placement (equivalent to one with preference value :1).
+/// * `yb_servers_refresh_interval` - Time interval, in seconds, between two attempts to refresh the information about cluster nodes.
+///     Default is 300. Valid values are integers between 0 and 600. Value 0 means refresh for each connection request. Any value
+///     outside this range is ignored and the default is used.
+/// * `fallback_to_topology_keys_only` - (default value: 5 seconds) The driver marks a server as failed with a timestamp, when it cannot
```
Comment (ashetkar) [inline tokio-postgres/src/config.rs]: Same here: descriptions got interchanged.

File: tokio-postgres/src/config.rs [hunk 6]
```diff
@@ -520,6 +564,141 @@ impl Config {
...
+    }
+
+    /// YugabyteDB Specific.
+    ///
+    /// Gets the failed_host_reconnect_delay_secs value.
+    pub fn get_failed_host_reconnect_delay_secs(&self) -> Duration {
+        self.failed_host_reconnect_delay_secs
+    }
+
+    ///Check is a given zone in Topology keys is valid
```
Comment (ashetkar) [inline tokio-postgres/src/config.rs]: nit: `Check is a given zone` -> `Check if a given zone`

File: tokio-postgres/src/connect.rs [hunk 7]
```diff
@@ -83,6 +132,433 @@ where
...
+where
+    T: MakeTlsConnect<Socket>,
+{
+    if config.host.is_empty() && config.hostaddr.is_empty() {
+        return Err(Error::config("both host and hostaddr are missing".into()));
+    }
+
+    if !config.host.is_empty()
+        && !config.hostaddr.is_empty()
+        && config.host.len() != config.hostaddr.len()
```
Comment (ashetkar) [inline tokio-postgres/src/connect.rs:148]: How are `host` and `hostaddr` different?

File: tokio-postgres/src/connect.rs [hunk 8]
```diff
@@ -83,6 +132,433 @@ where
...
+        }
+    }
+}
+
+fn increase_connection_count(host: Host) {
+    let mut conn_map = CONNECTION_COUNT_MAP.lock().unwrap();
+    let count = conn_map.get(&host);
+    if count.is_none() {
+        conn_map.insert(host.clone(), 1);
+        info!("Increasing connection count for {:?} to 1", host.clone());
```
Comment (ashetkar) [inline tokio-postgres/src/connect.rs]: Should be debug level log, if it's available.

File: tokio-postgres/src/connect.rs [hunk 9]
```diff
@@ -3,13 +3,62 @@ use crate::config::{Host, LoadBalanceHosts, TargetSessionAttrs};
...
+    };
+    static ref PUBLIC_HOST_MAP: Mutex<HashMap<Host, Host>> = {
+        let m = HashMap::new();
+        Mutex::new(m)
+    };
+    static ref HOST_TO_PORT_MAP: Mutex<HashMap<Host, u16>> = {
+        let m = HashMap::new();
+        Mutex::new(m)
+    };
+}
```
Comment (ashetkar) [inline tokio-postgres/src/connect.rs:58]: IMO, a struct would have helped keep these data structures together

File: tokio-postgres/src/connect.rs [hunk 10]
```diff
@@ -83,6 +132,433 @@ where
...
+    return true;
+}
+
+fn add_to_failed_host_list(host: Host) {
+    let mut failedhostlist = FAILED_HOSTS.lock().unwrap();
+    failedhostlist.insert(host.clone(), Instant::now());
+    info!("Added {:?} to failed host list", host.clone());
+}
+
+async fn refresh(client: Client, config: &Config) {
```
Comment (ashetkar) [inline tokio-postgres/src/connect.rs:444]: `refresh()` should be called by only one thread, if there are multiple threads trying to create connections. How is it ensured here?

File: tokio-postgres/src/connect.rs (hunk 7 above)
Comment (HarshDaryani896) [reply tokio-postgres/src/connect.rs:148]: It is explained here in the comments: https://github.com/yugabyte/rust-postgres/blob/master/postgres/src/config.rs#L39

Comment (HarshDaryani896) [reply tokio-postgres/src/connect.rs:148]: Discussed on call.

File: tokio-postgres/src/connect.rs (hunk 10 above)
Comment (HarshDaryani896) [reply tokio-postgres/src/connect.rs:444]: Fixed.

**Review by ashetkar — APPROVED** [review]

**Pull request merged**